--- a/mlx/backend/metal/kernels/gated_delta_update_nax.h
+++ b/mlx/backend/metal/kernels/gated_delta_update_nax.h
@@ -127,8 +127,8 @@
       gemm_op
           .template get_right_input_cooperative_tensor<AType, BType, CType>();
   auto ct_c = gemm_op.template get_destination_cooperative_tensor<
-      decltype(ct_a),
-      decltype(ct_b),
+      metal::remove_addrspace_t<decltype(ct_a)>,
+      metal::remove_addrspace_t<decltype(ct_b)>,
       CType>();
 
   STEEL_PRAGMA_UNROLL
@@ -173,8 +173,8 @@
       gemm_op
           .template get_right_input_cooperative_tensor<AType, BType, CType>();
   auto ct_c = gemm_op.template get_destination_cooperative_tensor<
-      decltype(ct_a),
-      decltype(ct_b),
+      metal::remove_addrspace_t<decltype(ct_a)>,
+      metal::remove_addrspace_t<decltype(ct_b)>,
       CType>();
 
   STEEL_PRAGMA_UNROLL
@@ -226,8 +226,8 @@
 
   // Create matmul output in register
   auto ct_c = gemm_op.template get_destination_cooperative_tensor<
-      decltype(ct_a),
-      decltype(ct_b),
+      metal::remove_addrspace_t<decltype(ct_a)>,
+      metal::remove_addrspace_t<decltype(ct_b)>,
       CType>();
 
   // Load A in to left operand registers
